File: Continuous Probability Distributions/Python/Pareto/Pareto.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import pareto

# A distribuição de Pareto é conhecida pelo "Princípio de Pareto" ou regra do 80/20.
# É amplamente utilizada para modelar a distribuição de renda, tamanho de cidades
# e ocorrências de sinistros em seguros. Ela possui suporte limitado à esquerda (x >= scale).

# Versão Distribuição de Pareto

# 1. Definição dos parâmetros da Distribuição de Pareto
shape = 3.0  # Parâmetro de forma (alfa) - Deve ser maior que 0
scale = 2.0  # Parâmetro de escala (xm / limite mínimo) - Deve ser maior que 0

# Determinar os limites de visualização baseados nos quantis (0% até 99.9%)
# O ponto inicial exato da distribuição é o valor do próprio parâmetro de escala
x_min = scale
x_max = pareto.ppf(0.999, b=shape, scale=scale)

# 2. Suporte: Valores contínuos iniciando a partir do limite mínimo (scale) até x_max
x = np.linspace(x_min, x_max, 1000)

# 3. Cálculo das funções exatas usando o SciPy nativo
# pareto.pdf equivale a dpareto (Densidade)
# pareto.cdf equivale a ppareto (Acumulada)
fx = pareto.pdf(x, b=shape, scale=scale)
Fx = pareto.cdf(x, b=shape, scale=scale)

# ==============================================================================
# Configuração da figura para os gráficos lado a lado (equivalente ao patchwork)
# ==============================================================================
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 5))

# Estilo visual limpo (equivalente ao theme_classic)
for ax in [ax1, ax2]:
    ax.spines['top'].set_visible(False)
    ax.spines['right'].set_visible(False)
    ax.tick_params(top=False, right=False)

# ------------------------------------------------------------------------------
# Gráfico 1: PDF (Função de Densidade de Probabilidade)
# ------------------------------------------------------------------------------
# Desenha a curva contínua da densidade (Decai rapidamente a partir do pico em x_min)
ax1.plot(x, fx, color="steelblue", linewidth=2)
# Preenche a área sob a curva para melhor visualização (alpha controla a transparência)
ax1.fill_between(x, fx, color="steelblue", alpha=0.15)

# Configurações de eixos e títulos dinâmicos
ax1.set_xlim(x_min, x_max)
ax1.set_ylim(0, max(fx) * 1.1)
ax1.set_title(f"Função de Densidade (PDF) - Pareto(shape={shape}, scale={scale})", weight='bold', pad=15)
ax1.set_xlabel("Valor (x)")
ax1.set_ylabel("f(x)")

# ------------------------------------------------------------------------------
# Gráfico 2: CDF (Função de Distribuição Acumulada Contínua)
# ------------------------------------------------------------------------------
# Desenha a curva contínua acumulada (Inicia exatamente no valor de scale)
ax2.plot(x, Fx, color="darkorange", linewidth=2)

# Configurações de eixos e títulos
ax2.set_xlim(x_min, x_max)
ax2.set_ylim(0, 1.05)
ax2.set_yticks(np.arange(0, 1.1, 0.2))  # Quebras de 0 até 1 de 0.2 em 0.2
ax2.set_title("Distribuição Acumulada (CDF)", weight='bold', pad=15)
ax2.set_xlabel("Valor (x)")
ax2.set_ylabel("F(x) = P(X <= x)")

# Ajusta o espaçamento para não cortar os textos
plt.tight_layout()

# Salvar como JPEG (Equivalente ao ggsave)
# plt.savefig("figPareto.jpeg", dpi=300, format="jpeg")

# Exibir os gráficos na tela (Equivalente ao x11() + plot)
plt.show()

File: Continuous Probability Distributions/R/Pareto/Pareto.R
# A distribuição de Pareto é conhecida pelo "Princípio de Pareto" ou regra do 80/20.
# É amplamente utilizada para modelar a distribuição de renda, tamanho de cidades
# e ocorrências de sinistros em seguros. Ela possui suporte limitado à esquerda (x >= scale).

# Versão Distribuição de Pareto

# Carregar as bibliotecas necessárias
if(!require(ggplot2)) install.packages("ggplot2"); library(ggplot2)
# Para juntar os gráficos lado a lado (equivalente ao plt.subplots)
if(!require(patchwork)) install.packages("patchwork"); library(patchwork)

# 1. Definição dos parâmetros da Distribuição de Pareto
shape <- 3.0  # Parâmetro de forma (alfa) - Deve ser maior que 0
scale <- 2.0  # Parâmetro de escala (xm / limite mínimo) - Deve ser maior que 0

# Determinar os limites de visualização baseados nos quantis (0% até 99.9%)
# O ponto inicial exato da distribuição é o valor do próprio parâmetro de escala
# Função quantil (inversa da CDF) exata para Pareto: xm / (1 - p)^(1 / alfa)
x_min <- scale
x_max <- scale / (1 - 0.999)^(1 / shape)

# 2. Suporte: Valores contínuos iniciando a partir do limite mínimo (scale) até x_max
x <- seq(x_min, x_max, length.out = 1000)

# 3. Cálculo das funções exatas usando fórmulas matemáticas analíticas
# PDF exata: (alfa * xm^alfa) / x^(alfa + 1)
pdf_pareto <- (shape * (scale^shape)) / (x^(shape + 1))

# CDF exata: 1 - (xm / x)^alfa
cdf_pareto <- 1 - (scale / x)^shape

dados_pareto <- data.frame(
  x  = x,
  fx = pdf_pareto,
  Fx = cdf_pareto
)

# ==============================================================================
# Gráfico da PDF (Função de Densidade de Probabilidade)
# ==============================================================================
g1 <- ggplot(dados_pareto, aes(x = x, y = fx)) +
  # Desenha a curva contínua da densidade (Decai rapidamente a partir do pico em x_min)
  geom_line(color = "steelblue", linewidth = 1) +
  # Preenche a área sob a curva para melhor visualização (alpha controla a transparência)
  geom_area(fill = "steelblue", alpha = 0.15) +
  # Configurações de eixos e títulos dinâmicos
  scale_x_continuous(expand = c(0, 0), limits = c(x_min, x_max)) +
  scale_y_continuous(limits = c(0, max(dados_pareto$fx) * 1.1), expand = c(0, 0)) +
  labs(
    title = paste0("Função de Densidade (PDF) - Pareto(shape=", shape, ", scale=", scale, ")"),
    x = "Valor (x)",
    y = "f(x)"
  ) +
  theme_classic() + # Remove as bordas superior e direita (equivalente às spines ocultas)
  theme(plot.title = element_text(hjust = 0.5, face = "bold"))

# ==============================================================================
# Gráfico da CDF (Função de Distribuição Acumulada Contínua)
# ==============================================================================
g2 <- ggplot(dados_pareto, aes(x = x, y = Fx)) +
  # Desenha a curva contínua acumulada (Inicia exatamente no valor de scale)
  geom_line(color = "darkorange", linewidth = 1) +
  # Configurações de eixos e títulos (seq equivale ao np.arange)
  scale_x_continuous(expand = c(0, 0), limits = c(x_min, x_max)) +
  scale_y_continuous(limits = c(0, 1.05), breaks = seq(0, 1, 0.2), expand = c(0, 0)) +
  labs(
    title = "Distribuição Acumulada (CDF)",
    x = "Valor (x)",
    y = "F(x) = P(X <= x)"
  ) +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5, face = "bold"))

# ==============================================================================
# Exibir e salvar os gráficos lado a lado
# ==============================================================================
grafico_final <- g1 + g2

# Abre a janela gráfica (Equivalente ao plt.show())
x11(width = 12, height = 5)
grafico_final

# Salvar como JPEG (Equivalente ao plt.savefig)
# ggsave("figPareto.jpeg", plot = grafico_final, width = 12, height = 5, dpi = 300)
